### PR TITLE
Make GenericDocument attributes protected

### PR DESCRIPTION
**Pull request by psarna**
In order to be able to inherit from GenericDocument and use its
private variables (most importantly - have access to stack_),
the attributes are now declared protected instead of private.

One future use case (for ScyllaDB) involves creating a subclass
for GenericDocument, which overrides its Handler interface
(StartObject, EndObject, etc.), adding additional checks to prevent
too deeply nested JSON objects from being created.

Signed-off-by: Piotr Sarna <sarna@scylladb.com>

### CODE REVIEW

**Comment by miloyip** [comment]
I am not sure if this is an good idea.
Alternatively, to do additional checks, it is possible to use `Document::Populate(checker)`, where checker is an object which delegates the events from a reader and do additional checking.

**Comment by psarna** [comment]
Alright, but with `Populate(checker)`, the document will be traversed twice, once for validation and once for actual parsing - which can be undesirable for performance. Let me just explain the experience that led me to create this pull request:
1. Since the `GenericDocument` class is not marked `final`, it's fine to inherit from it, which I did, overriding some of its methods (`StartObject`, `EndObject` and some other ones).
2. `GenericDocument::Parse()` uses `GenericReader::Parse` and passes itself by reference to it, which was not good enough for my concept, since it will pass a reference to `GenericDocument &`, not my subclass.
3. Fortunately, I can also use the `GenericReader::Parse` directly - which works *almost* like a charm.
4. The only problem is that after calling `GenericReader::Parse(stream, my_document)` I have no easy way to extract the parsed document out of my subclass... it's kept inside `GenericDocument::stack_`, which my subclass has no access to - because the field is private instead of protected.

Currently I'm fine with using a workaround - after calling `GenericReader::Parse()`, I call `Populate()` with an empty generator, which does nothing. But, `Populate()` also has a side effect of popping the top element from the internal `GenericDocument::stack_`, which is exactly what I need - after all, the stack already has the document parsed by `GenericReader::Parse` I called earlier.

So, to sum up, I have everything I need to parse the document without traversing it twice, but it's a little bit hacky. With `stack_` being protected instead of private, I could simply pop it and get the value I need. It just feels like this field should not be private if inheriting from `GenericDocument` is allowed.

I don't insist though, since as I already mentioned, I have functioning workaround for my issue - feel free to close this pull request if you prefer not to have it merged.

Thanks!

**Comment by miloyip** [comment]
You may check http://rapidjson.org/md_doc_schema.html#DOM , which does schema validation while parsing a stream into a document. If it finds invalid situation during the parsing, it can stop immediately as well. In this example, it does not have additional traversal.

**Comment by sdegrande** [comment]
Hi.

I mostly have the same need.

From big GeoJSon files, we 'extract' objects  (called "Feature") which have a given "type" member value, and we construct our own data structure based on those objects. We would like to avoid to have the whole DOM in memory (which could be very big).

We need to parse a whole "Feature" object (which contains several sub-objects, arrays...) before to add it into our structure. Using the SAX API is thus a pain, because we have to handle a stack, so we are more or less re-implementing a Document.

We just need to be called on EndObject(), and so we use the same "trick" than @psarna : inherit from Document, implement our EndObject(), and our ParseStream().

So, now the idea is to avoid to fill the memory, and so in our EndObject(), we call Document::EndObject(), get the top of the stack, check if it is a "Feature", if so construct our data struct and call SetNull() on the Value from the top of the stack to release the now unneeded Value from memory.

So we only need a way to get access to the top of the stack. The proposed patch from @psarna or perhaps just a function to get the stack's top would be perfect !
